### DISCUSSION TITLE
How can we set up a dev environment for step debugging Conduit Connectors?

### DISCUSSION THREAD

**Discussion by nickchomey** [ConduitIO/conduit#1724] [Q&A] [answered]
I'd like to make a connector of my own and perhaps even contribute to some existing ones, but I dont know how to step debug the connectors given that you run the conduit binary, which fetches/includes the connector binaries. 

Could anyone please provide some tips?

In fact, I think this would be a very good topic to make an official doc for.

**Comment by hariso** [comment]
>  which fetches/includes the connector binaries

Can you clarify this part please?

But generally, for built-in (included) plugins, you can use your IDE and simply put breakpoints. As for the standalone plugins: they are, essentially, just compiled that are being run, so you can use remote debugging for them. The instructions for that depend on the IDE.

**Reply by nickchomey** [reply]
Sorry, Im quite sure that I just wrote clumsily.

Yes, I'm talking about standalone plugins. I can generally step debug a "primary" binary, but don't know how to step debug when a binary executes another binary, such as with the standalone plugins.

I use vs code, but I think that most ides use delve. If you could at least point me in the rejght direction of some documentation on how to do this, it would be very appreciated. But I do think it would be valuable to document setting up a basic dev environment for, say, vscode and maybe goland. It would allow people get up to speed more quickly (or at all), and be able to contirbute to conduit!

Thanks!

**Reply by lovromazgon** [reply]
@nickchomey you might have missed [my answer](https://github.com/ConduitIO/conduit/discussions/1724#discussioncomment-10178484), let me know if that's what you were looking for.

**Comment by lovromazgon** [comment]
I've debugged standalone plugins in the past, I've quickly compiled the approach I used. Note that this is a bit experimental, when I tried it out now, the plugin didn't finish gracefully for some reason and the Conduit process had to be killed. Nevertheless, debugging itself should work.

1. Install delve.

Full guide: https://github.com/go-delve/delve/tree/master/Documentation/installation

TL;DR:

```sh
go install https://github.com/go-delve/delve@latest
```

2. Wrap connector binary in a script that runs it using delve.

For this I suggest moving the standalone connector binary into a sub-folder in `connectors`, so Conduit doesn't load it directly (it only loads binaries it finds directly in the `connectors` folder). You should end up with something like this:

```
my-conduit-path
├── conduit                        # conduit binary
└── connectors
    ├── bin
    │   └── conduit-connector-file # connector binary
    └── conduit-connector-file.sh  # entrypoint that wraps connector in delve call
```

The contents of `conduit-connector-file.sh`:

```sh
#!/usr/bin/env bash
BASEDIR=$(dirname "$0")
dlv --listen=:9876 --headless=true --api-version=2 --accept-multiclient --log-dest=2 exec "$BASEDIR/bin/conduit-connector-file"
```

This will run delve on port 9876 and prepare it for remote debugging.

3. Run Conduit and debug the plugin

When starting Conduit, it first scans the `connectors` folder, runs each connector and fetches its specifications. The connector will block though, because delve is waiting for a remote debugger to attach and continue the execution. So right after starting Conduit you need to attach to the plugin and unblock it:

```sh
$ dlv connect :9876
(dlv) continue
(dlv) exit
Remote process has exited. Would you like to kill the headless instance? [Y/n] y
```

Assuming you have a pipeline configuration file that will use this standalone plugin, you need to attach a second time to actually debug it while the pipeline is running (so again `dlv connect :9876` like above). Then you can set any breakpoints you wish. If you prefer, you can also use an IDE to do this instead of the CLI interface.

Keep in mind that Conduit waits for a limited amount of time for the plugin to start and do the handshake (I think 1 minute?), after that it will consider it broken and detach from the process, so you need to be relatively quick to attach to the plugin and advance execution.

**Reply by nickchomey** [reply]
This looks great, thanks very much! I didn't see it previously as I was just using the Github android app, which loads threads rather than the whole issue.

I'll use this as a starting point and fiddle around. If I get it working, I'll report back with instructions.

Thanks again!

**Reply by nickchomey** [reply]
Here are some clarifications that will probably help others to follow this:

1. Add a new section to the connector's Makefile
```bash
build-debug:
	go build -gcflags=all="-N -l" -ldflags="-X 'github.com/conduitio-labs/conduit-connector-surrealdb.version=${VERSION}'" -o conduit-connector-surrealdb cmd/connector/main.go
```
The `-gcflags=all="-N -l"` , allow for building the binary in a way that ensures proper step debugging. run `make build-debug` and put that binary in the `connectors/bin` directory mentioned in the above instructions

2. If you want to step debug in vscode, create a launch.json config like this

```json
{
            "name": "Attach to Conduit Connector",
            "type": "go",
            "request": "attach",
            "mode": "remote",
            "cwd": "${workspaceFolder}/conduitproject/connectors/conduit-connector-nats-pubsub", // This points to the connector's directory where you ran `make build` from. I found that `cwd` was needed, but `"remotePath"` wasn't.. YMMV.
            "port": 9876,
            "host": "127.0.0.1",
            "showLog": true,
            "logOutput": "rpc,debuglineerr,dap,debugger", //select these as-desired
            "trace": "verbose",
            "debugAdapter": "dlv-dap", // This is a new adapter, but the default setting was reverted to "legacy" while they work on some issues. But I find that "dlv-dap" works fine here and has some important improvements over legacy
}
```
3. Set breakpoints in your connector's code - perhaps in the Open and Write methods. 
4. You do not need to run the bash script manually. Instead, just run the conduit binary (e.g. `./conduit` and it will automatically run the bash script if it is in the main `/connectors` directory. When you do that, it will launch conduit and pause - only outputting something like this
![image](https://github.com/user-attachments/assets/19507a67-3438-4316-91c1-35f559b1800e)

5. Open another terminal and run the commands from above. The rest of the normal conduit logging will fill the terminal, showing that the connectors have successfully started etc... 
```
$ dlv connect :9876
Type 'help' for list of commands.
(dlv) continue
Process 273701 has exited with status 0
(dlv) exit
Remote process has exited. Would you like to kill the headless instance? [Y/n] y
```
6. Now run the step debugger - Press F5, or select it from the dropdown if needed
![image](https://github.com/user-attachments/assets/4b316653-364f-44d6-aeec-16f7ef4d745d)

It should just proceed and trigger the breakpoints now. Step through them as you would normally. 

7. I found that you cant kill conduit as you normally would (Ctrl+C). I found success with `pkill -f conduit`. And perhaps also `pkill -f dlv` if you find that any terminal is still stalled.

**Reply by nickchomey** [reply]
We can also debug numerous connectors (eg. source and destination) at the same time by duplicating the same process as above. 
1. Create another bash script and change the dlv port to something else - e.g. 9877 (the number doesnt matter). 
2. Move the connector to the bin subdirectory
3. Create a new entry in launch.json and change the port to 9877 and path to the connector's source code root directory
4. Launch conduit
5. do the `dlv connect->continue->exit` for the first one. 
6. Then do `dlv` again for the second, changing the port number
7. attach the first vs code debugger. 
8. Attach the 2nd vs code debugger. 

Both debuggers will be activated, and it will hit the first breakpoint in the source connector first, and you can step through as-needed. Then when done, it'll hit the first breakpoint in the destination connector.

**Reply by nickchomey** [reply]
I've submitted a PR to add a debug build, but nothing else. I don't know what to do to put all of the above in a make file - you have to have at least 2 terminals, unless there's a way to start conduit in the background somehow? You're welcome to suggest/submit improvements 

https://github.com/ConduitIO/conduit-connector-template/pull/48

**Reply by hariso** [reply]
I don't think we need to do anything special about Conduit, because it's easy to debug an application you yourself are running. I guess you can just run Conduit in debug mode using VS Code. The problem with connectors is that Conduit is running them when needed, so you don't know when to attach to the connector processes.

**Reply by nickchomey** [reply]
I actually was not able to debug conduit itself yesterday. I followed the same process as for the connectors bug it kept freezing with some warning about debugging optimized code (even though I built it with the non-optimized flags).

I suppose I could try just launching a debug process for conduit without building a binary - as you normally would for debugging go applications - but I'm not sure how to load connectors when doing that

Any suggestions?

**Reply by hariso** [reply]
I'm using GoLand, and there I just use the debug option. No special setup at all needed, GoLand does everything. Nothing should change for connector, you still can use the same approach as above to debug connectors.

**Comment by nickchomey** [comment, marked as answer]
It was later suggested that we just include connectors we're developing as built-in connectors rather than standalone. This avoids all of the difficulties with Delve etc... 

The docs are straightforward: https://conduit.io/docs/connectors/additional-built-in-plugins/

For my use-case of mysql and surrealdb you just do the following:

```
package main

import (
	"github.com/conduitio/conduit/pkg/conduit"
	mysql "github.com/conduitio-labs/conduit-connector-mysql"
	surrealdb "github.com/nickchomey/conduit-connector-surrealdb"
)

func main() {
	// Get the default configuration, including all built-in connectors
	cfg := conduit.DefaultConfig()

	cfg.ConnectorPlugins["mysql"] = mysql.Connector
	cfg.ConnectorPlugins["surrealdb"] = surrealdb.Connector


	conduit.Serve(cfg)
}
```

If you want to step debug the local dev version of these packages, just use `replace`

```
replace (
	github.com/conduitio-labs/conduit-connector-mysql v0.1.0 => ~/conduitproject/connectors/conduit-connector-mysql
	github.com/conduitio/conduit v0.12.0 => ~/conduitproject/conduit
	github.com/nickchomey/conduit-connector-surrealdb v0.1.0 => ~/conduitproject/connectors/conduit-connector-surrealdb
)
```
You can then add breakpoints to this main.go file, the core conduit files, or the connector files. And just launch main.go with VS Code's debugger with a launch.json config like this (feel free to change `${file}` to a specific path etc... 

```
        {
            "name": "Launch current main.go",
            "type": "go",
            "request": "launch",
            "mode": "auto",
            "program": "${file}",
            "showLog": true,
            "env": {},
            "args": [],
            "trace": "verbose",
            "debugAdapter": "dlv-dap"
        },
```
